File: backend/playback.py
import time

from config import ADVANCE_DEDUP_WINDOW
from models import Room


def go_next(room: Room) -> bool:
    """Advance to next track, removing the finished one. Returns True if position changed."""
    if not room.queue:
        return False

    # A client "ended" event and the server-side auto-advance loop (or several
    # clients at once) can all fire within a few milliseconds. Ignore repeats
    # inside a short window so the queue only moves forward by one.
    now = time.time()
    if now - room.last_advance_at < ADVANCE_DEDUP_WINDOW:
        return False
    room.last_advance_at = now

    # current_index can drift out of range if the queue shrank elsewhere;
    # clamp before popping so we never raise IndexError.
    idx = max(0, min(room.current_index, len(room.queue) - 1))
    finished = room.queue[idx]
    if finished.source_url:
        room.radio_seed_url = finished.source_url
    room.queue.pop(idx)
    room.current_index = max(0, min(idx, len(room.queue) - 1))
    room.position = 0

    if not room.queue:
        room.is_playing = False
        return True

    room.is_playing = True
    room.last_sync_at = time.time()
    return True


def go_prev(room: Room) -> bool:
    """Go to previous track. Returns True if position changed."""
    if not room.queue or room.current_index <= 0:
        return False
    room.current_index -= 1
    room.position = 0
    room.is_playing = True
    room.last_sync_at = time.time()
    return True


def jump_to(room: Room, index) -> bool:
    """Jump to a specific track by index. Returns True if successful."""
    if not isinstance(index, (int, float)) or index != int(index):
        return False
    index = int(index)
    if 0 <= index < len(room.queue):
        room.current_index = index
        room.position = 0
        room.is_playing = True
        room.last_sync_at = time.time()
        return True
    return False


def seek_to(room: Room, position: float) -> None:
    """Seek to a specific position in the current track."""
    room.position = max(0.0, position)
    room.last_sync_at = time.time()

File: backend/internal/domain/playback.go
package domain

import "sort"

func sortStrings(s []string) { sort.Strings(s) }

// GoNext advances to the next track, dropping the one that just finished, and
// reports whether the room's position actually moved.
//
// dedupWindow guards the queue against double-skips: a client "ended" event and
// the server-side auto-advance loop (or several clients at once) can all fire
// within a few milliseconds, and only the first should count.
func (r *Room) GoNext(dedupWindow float64) bool {
	if len(r.Queue) == 0 {
		return false
	}

	now := Now()
	if now-r.LastAdvanceAt < dedupWindow {
		return false
	}
	r.LastAdvanceAt = now

	// CurrentIndex can drift out of range if the queue shrank elsewhere; clamp
	// before removing so we never index past the end.
	idx := r.CurrentIndex
	if idx < 0 {
		idx = 0
	}
	if idx > len(r.Queue)-1 {
		idx = len(r.Queue) - 1
	}

	finished := r.Queue[idx]
	if finished.SourceURL != "" {
		r.RadioSeedURL = finished.SourceURL
	}

	r.Queue = append(r.Queue[:idx], r.Queue[idx+1:]...)
	r.CurrentIndex = clampIndex(idx, len(r.Queue))
	r.Position = 0

	if len(r.Queue) == 0 {
		r.IsPlaying = false
		return true
	}

	r.IsPlaying = true
	r.LastSyncAt = Now()
	return true
}

// GoPrev steps back one track and reports whether anything changed.
func (r *Room) GoPrev() bool {
	if r.CurrentIndex <= 0 {
		return false
	}
	r.CurrentIndex--
	r.Position = 0
	r.IsPlaying = true
	r.LastSyncAt = Now()
	return true
}

// JumpTo moves to a specific queue position and reports whether the index was
// in range.
func (r *Room) JumpTo(index int) bool {
	if index < 0 || index >= len(r.Queue) {
		return false
	}
	r.CurrentIndex = index
	r.Position = 0
	r.IsPlaying = true
	r.LastSyncAt = Now()
	return true
}

// SeekTo moves playback to a position within the current track and re-anchors
// the sync point, so every client extrapolates from the same instant.
func (r *Room) SeekTo(position float64) {
	r.Position = position
	r.LastSyncAt = Now()
}

// Enqueue appends a track and starts playback if the room was idle with an
// empty queue.
func (r *Room) Enqueue(track *Track) {
	r.Queue = append(r.Queue, track)
	if track.SourceURL != "" {
		r.RadioSeedURL = track.SourceURL
	}
	if len(r.Queue) == 1 {
		r.IsPlaying = true
		r.Position = 0
		r.LastSyncAt = Now()
	}
}

func clampIndex(idx, length int) int {
	if length <= 0 {
		return 0
	}
	if idx < 0 {
		return 0
	}
	if idx > length-1 {
		return length - 1
	}
	return idx
}
